### ISSUE TITLE
Join not producing the expected SQL

### ISSUE THREAD

**Issue by andre-artus** [requery/requery#392]
I'm trying to get a simple join to work, but there is either a bug or I'm missing something:

``` Kotlin
// in Kotlin
  val assets = data
                .select(AssetEntity.BAR_CODE, AssetDescriptionEntity.DESCRIPTION)
                .from(AssetEntity::class)
                .join(AssetDescriptionEntity::class)
                .on(AssetEntity.DESCRIPTION_ID.eq(AssetDescriptionEntity.ID))
                .where(AssetEntity.ROOM_CODE.eq(barCode))
                .get()
                .toList()
```
produces the following 
`select a.bar_code, b.description from assets a inner join descriptions b on a.descriptionId = ? where a.roomCode = ?`
Which returns no results, what I want is the following:

```
SELECT
  a.bar_code,
  b.description
FROM assets a
  INNER JOIN descriptions b ON a.descriptionId = b.id
WHERE a.roomCode = ?
```

Also, is there a way I can set up an entity so that I can write the foreign key but read the related entity?
E.g.
```

interface Bar {
    var fooCode: String  // write to this
    var foo: Foo // read from this
}

interface Foo {
    @get:Key
    var code: String

    @get:OneToMany(mappedBy = "foo")
    val bars: MutableSet<Bar>
}
```

**Comment by Agraphie** [comment]
I tried to reproduce your first example, but you did not post the entity classe. I tried your query with the following (guessed) classes

```Java
@Entity
@Table(name = "assets")
public interface Asset {
  @Key
  @Generated
  int getId();

  String getBarCode();

  @OneToOne
  @ForeignKey
  AssetDescription getDescription();

  int getRoomCode();
}
```

```Java
@Entity
@Table(name = "descriptions")
public interface AssetDescription {
  @Key @Generated
  int getId();

  String getDescription();
}
```

And executing the query you posted results in this query for me (which is the one you wanted)
`select a.barCode, b.description from assets a inner join descriptions b on a.description = b.id where a.roomCode = ? `
Maybe you could post your entity classes?

For the second question, I'm not entirely sure what you mean.

**Comment by andre-artus** [comment]
```
@Entity
@Table(name = "descriptions")
interface AssetDescription : Parcelable, Persistable {
    @get:Key
    val id: Int

    @get:Column(unique = true)
    var description: String

    @get:OneToMany(mappedBy = "descriptionId")
    val assets: MutableSet<Asset>
}

@Entity
@Table(name = "assets")
interface Asset : Parcelable, Persistable {
    @get:Key
    @get:Column(name = "bar_code", nullable = false)
    var barCode: String

    @get:ForeignKey(references = Room::class)
    var roomCode: String?
  
    @get:ForeignKey(references = AssetDescription::class)
    var descriptionId: Int
}

```
None of the keys are auto generated they come from an external source.

For the second question: 

When creating a new `asset `I would like to be able to set the `description `on the `asset `by just assigning to `descriptionId `and not have to load a description object into memory just to then assign it, and that is why I'm using `descriptionId` instead of `AssetDescription getDescription();`. It would be nice to have both so that I can assign to `descriptionId` when creating the `asset` but read from `getDescription();` later on (As I can do in Entity Framework) The initial data is loaded as a batch and creating the extra objects has a tangible effect on how long it takes to complete.

Thank you.

**Comment by npurushe** [comment]
Hi currently you can only define it one way, either placing the id as field in the entity, or as a entity reference. Both are supported but not both at the same time as keeping them in sync would be a problem e.g. if you changed descriptionId then the entity being referenced to would now be incorrect something that would have to be accounted for.

The first issue just may only be a problem in the kotlin library looking into it, thanks

**Comment by andre-artus** [comment]
I thought as much, npurushe, but thought to ask just in case. For my use case populating the database and using it are distinct so that I would not need for them to be in sync, but I realize that's not the case for the majority of users. There is probably a bunch of notifications sent to facilitate binding and so on.

**Comment by andre-artus** [comment]
Is there a way to do a raw SQL query in requery when using Kotlin? I see mentions of data.raw() in the issues (including the issue just after mine), but it does not seem to be available from KotlinEntityDataStore.

**Comment by npurushe** [comment]
That was missing it was added into master for the next update

**Comment by npurushe** [comment]
Thanks for reporting this fix is available in 1.1.1
